src/pudl/transform/ferc1.py

```diff
@@ -124,6 +124,117 @@ class Config:
+
+    Reads all rows that need to be mapped out of the ``f1_row_lit_tbl`` and
+    appends columns containing any previously mapped values, returning the
+    resulting dataframe.
+
+    This
+    """
+    idx_cols = ["sched_table_name", "row_status", "row_number", "report_year"]
+    all_rows = get_ferc1_dbf_rows_to_map(ferc1_engine).set_index(idx_cols)
+    with importlib.resources.open_text(
+        "pudl.package_data.ferc1", "dbf_to_xbrl.csv"
+    ) as file:
+        mapped_rows = (
+            pd.read_csv(file).set_index(idx_cols).drop(["row_literal"], axis="columns")
+        )
+    return (
+        pd.concat([all_rows, mapped_rows], axis="columns")
+        .reset_index()
+        .sort_values(["sched_table_name", "row_status", "report_year", "row_number"])
+    )
+
+
+def get_ferc1_dbf_xbrl_glue(dbf_table_name: str) -> pd.DataFrame:
+    """Read the DBF+XBRL glue and expand it to cover all DBF years."""
+    with importlib.resources.open_text(
+        "pudl.package_data.ferc1", "dbf_to_xbrl.csv"
+    ) as file:
+        row_map = pd.read_csv(
+            file,
+            usecols=[
+                "sched_table_name",
+                "row_status",
+                "report_year",
+                "row_number",
+                "row_type",
+                "xbrl_column_stem",
+            ],
+        )
+    # Select only the rows that pertain to dbf_table_name
+    row_map = row_map.loc[
+        row_map.sched_table_name == dbf_table_name,
+        ["row_status", "report_year", "row_number", "row_type", "xbrl_column_stem"],
+    ]
+    # Indicate which rows have unmappable headers in them, to differentiate them from
+    # null values in the exhaustive index we create below:
+    row_map.loc[
+        (row_map.row_type == "header") & (row_map.xbrl_column_stem.isna()),
+        "xbrl_column_stem",
+    ] == "HEADER_ROW"
+    row_map = row_map.drop(["row_type"], axis="columns")
+
+    # Create an index containing all possible index values:
+    idx = pd.MultiIndex.from_product(
+        [
+            row_map.row_status.unique(),
+            Ferc1Settings().dbf_years,
+            row_map.row_number.unique(),
+        ],
+        names=["row_status", "report_year", "row_number"],
+    )
+
+    # Concatenate the row map with the empty index, so we have blank spaces to fill:
+    row_map = pd.concat(
+        [
+            pd.DataFrame(index=idx),
+            row_map.set_index(["row_status", "report_year", "row_number"]),
+        ],
+        axis="columns",
+    ).reset_index()
+
+    # Forward fill missing XBRL column names, until a new definition for the row
+    # number is encountered:
+    row_map["xbrl_column_stem"] = row_map.groupby(
+        "row_number"
+    ).xbrl_column_stem.transform("ffill")
+    # Drop any rows that do not actually map between DBF rows and XBRL columns:
+    row_map = (
+        row_map.replace({"HEADER_ROW": np.nan})
+        .dropna(subset=["xbrl_column_stem"])
```

Comment:
Ah yeah, I agree specifying the column is preferable.

The reason I'm using this label / sentinel value is so that we can distinguish between a row where `xbrl_column_stem` is NULL because it's between two valid mapped values and should thus be filled in, and a row that was NULL because it was a header row, which means it shouldn't be filled in. Without this, when a row number associated with a non-header row becomes a header-row due to a change in the meaning of the row numbers, the `ffill` above keeps right on filling and associates all of the new header rows with the value of `xbrl_column_stem` that was associated with the old row number.

Reply:
I've edited my explanation above a little and inserted it as a comment.